internal/adminapi/client.go

```diff
@@ -229,10 +239,22 @@ func (cf ClientFactory) CreateAdminAPIClient(ctx context.Context, discoveredAdmi
 	opts := cf.opts
 	opts.TLSServerName = discoveredAdminAPI.TLSServerName
 
-	cl, err := NewKongClientForWorkspace(ctx, discoveredAdminAPI.Address, cf.workspace, opts, cf.adminToken)
+	var client *Client
+	err := retry.Do(
```

Comment:
Do we need to retry in the factory itself? We already have retries in the `adminAPIClients` both for the service-discovery case and the static list of URLs. That would be another layer of retries, making the total retries count squared (`KongAdminInitializationRetries^2`).

Reply:
The retries in `adminAPIClients` here is used for creating admin clients not using gateway discovery (providing admin API URLs). It does not use the factory.

In the case for gatway discovery, `AdminAPIClientFromServiceDiscovery` is called and here it retries for getting the endpoint list, then create client for each Kong admin API endpoint by factory. This does not import another layer of retry.
(But separating the creation of clients in two places, one in factory but another not is not very clear and readable. A refactor may be required.)

Reply:
> The retries in adminAPIClients here is used for creating admin clients not using gateway discovery (providing admin API URLs). It does not use the factory.

You're right, I saw `NewKongClientForWorkspace` and wrongly assumed it was a factory call.

Shouldn't we adjust the retry [here](https://github.com/Kong/kubernetes-ingress-controller/blob/ea72f4aa3fc491d9cbb716c37b31ca519474da02/internal/manager/setup.go#L407) then to respect the existing knobs to fix it instead of adding retries to the factory?

Reply:
Moved the retries from factory to `AdminAPIClientFromServiceDiscovery`.